TheFunctionChangedMyMind/README-en.md

```diff
@@ -0,0 +1,322 @@
+
+## Reified for your types
+
+We are currently having a generic – `<T>` – function that is _inlined_, these two elements can be combined to use a third element on our function: `reified`:
+
+```kotlin
+inline fun <reified Ty> Context.launchActivity(Class<T> clazz, confBlock: Intent.() -> Unit) {
+    val intent = Intent(this, clazz)
+    intent.confBlock()
+    startActivity(intent)
+}
+```
+
+Adding this word will means that for every single call of our function, when inlining, the compiler will use the actual type instead of a generic or a super type in the caller. This small change allows us to use the generic inside of our function as a real type, then we can remove the first parameter and replace with the real type:
+
+```kotlin
+inline fun <reified T> Context.launchActivity(confBlock: Intent.() -> Unit) {
+    val intent = Intent(this, T::class.java)
+    intent.confBlock()
+    startActivity(intent)
+}
+```
+
+Now all the calls to this function can be as clean as:
+
+```kotlin
+launchActivity<SomeActivity> {
+    putExtra(MY_KEY, myValue)
+}
+
+launchActivity<MyNextActivity> {}
+```
+
+Some final touches to the signature will make the function even safer and even cooler when being called:
+
+1. Make `T` have a constraint to `Activity` in a way only classes that extend from `Activity` are allowed
+2. Add a default value to the _lambda with reciever_ to allow calls with empty configuration
+
+```kotlin
+inline fun <reified T: Activity> Context.launchActivity(confBlock: Intent.() -> Unit = {}) {
+    ...
+}
+```
+
+Now our function is cooler than ever, won't add too much to the final bytecode and also will make our code more readable. Wow! We traveled a long way doing a lot of changes to a single piece of code to make things cooler, but not just that, now other people developing along with us can benefit from this improvement as they can simply call our function without having to worry how it works. I have seen even people adding animations, making a way to hold results and transforming data using this kind of functions/approaches. Writting idiomatic Kotlin actually makes you think and develop different and cooler!
+
+## Conlusion
+
+We can apply this kind of approch by steps or go directly to other functions on our codebase, one example is the usage of `SharedPreferences` and making the _edition_ do an _auto-apply_ "magically" using the lambda with reciever. We can create ways to configure more stuff in less steps and this will make our writting of code improving as we are adding tools to our set of _already-vaste_ set of functions and tools on kotlin. Thinking on what is repeated or what can be improved allows us –developers– to discover new worlds and make things cooler and sharper. Don't be scared of the change.
```

Comment:
```suggestion
We can apply this kind of approch by steps or go directly to other functions on our codebase, one example is the usage of `SharedPreferences` and making the _edition_ do an _auto-apply_ "magically" using the lambda with receiver. We can create ways to configure more stuff in less steps and this will make our writing of code improve as we will be adding tools to our _already-vast_ set of functions and tools on Kotlin. Thinking about what is repeated or can be improved allows us –the developers– to discover new worlds and make things cooler and sharper. Don't be scared of change.
```